README.md
# toshi-hazard-utils


[![pypi](https://img.shields.io/pypi/v/toshi-hazard-utils.svg)](https://pypi.org/project/toshi-hazard-utils/)
[![python](https://img.shields.io/pypi/pyversions/toshi-hazard-utils.svg)](https://pypi.org/project/toshi-hazard-utils/)
[![Build Status](https://github.com/gns-science/toshi-hazard-utils/actions/workflows/dev.yml/badge.svg)](https://github.com/gns-science/toshi-hazard-utils/actions/workflows/dev.yml)
[![codecov](https://codecov.io/gh/gns-science/toshi-hazard-utils/branch/main/graphs/badge.svg)](https://codecov.io/github/gns-science/toshi-hazard-utils)



Helpers for the toshi-hazard-store


* Documentation: <https://gns-science.github.io/toshi-hazard-utils>
* GitHub: <https://github.com/gns-science/toshi-hazard-utils>
* PyPI: <https://pypi.org/project/toshi-hazard-utils/>
* Free software: GPL-3.0-only


## Features

* TODO

## Credits

This package was created with [Cookiecutter](https://github.com/audreyr/cookiecutter) and the [waynerv/cookiecutter-pypackage](https://github.com/waynerv/cookiecutter-pypackage) project template.


CHANGELOG.md
# Changelog

## 0.1.0 (2022-08-01)

* First release on PyPI.


tests/__init__.py
"""Unit test package for toshi-hazard-utils."""


tests/test_toshi_hazard_utils.py
#!/usr/bin/env python
"""Tests for `toshi_hazard_utils` package."""

import pytest
from click.testing import CliRunner

from toshi_hazard_utils import cli


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string
    del response


def test_command_line_interface():
    """Test the CLI."""
    runner = CliRunner()
    result = runner.invoke(cli.thu)
    assert result.exit_code == 0
    assert 'thu' in result.output
    help_result = runner.invoke(cli.thu, ['--help'])
    assert help_result.exit_code == 0
    assert '--help  Show this message and exit.' in help_result.output


tests/features/test_main_features.py
#!/usr/bin/env python
"""Tests for `toshi_hazard_utils` package."""

# import pytest
import unittest
import random
import itertools
import io
import csv

from moto import mock_dynamodb

from nzshm_common.grids.region_grid import load_grid
from nzshm_common.location.code_location import CodedLocation
from toshi_hazard_utils.hazard import hazard_report, ToshiHazardReport

from toshi_hazard_store import model, query_v3

HAZARD_MODEL_ID = "MYHAZID"
GRID_02 = load_grid('NZ_0_2_NB_1_1')
VS30S = [250, 350, 450]
IMTS = ['PGA', 'SA(0.5)']
AGGS = ['mean', '0.10']
LOCS = [CodedLocation(*o, 0.001) for o in GRID_02[20:50]]
N_LVLS = 29


def build_hazard_aggregation_models():

    lvps = list(map(lambda x: model.LevelValuePairAttribute(lvl=x / 1e3, val=(x / 1e6)), range(1, N_LVLS)))
    for (loc, vs30, agg) in itertools.product(LOCS, VS30S, AGGS):
        for imt, val in enumerate(IMTS):
            yield model.HazardAggregation(
                values=lvps,
                vs30=vs30,
                agg=agg,
                imt=val,
                hazard_model_id=HAZARD_MODEL_ID,
            ).set_location(loc)


@mock_dynamodb
class HighLevelHazard(unittest.TestCase):
    def setUp(self):
        model.migrate()
        with model.HazardAggregation.batch_write() as batch:
            for item in build_hazard_aggregation_models():
                batch.save(item)

    def test_list_hazard_data_available_for_grid_location(self):
        loc = LOCS[0]
        hazard_reports = list(hazard_report([HAZARD_MODEL_ID], locations=[loc.code]))
        print(hazard_reports)

        result = hazard_reports[0]
        self.assertTrue(isinstance(result, ToshiHazardReport))
        self.assertEqual(len(hazard_reports), 1)
        self.assertEqual(result.loc, loc)
        self.assertEqual(set(IMTS), result.imts)
        self.assertEqual(set(AGGS), result.aggs)
        self.assertEqual(set(VS30S), result.vs30s)
        self.assertEqual(set(lvl / 1e3 for lvl in range(1, N_LVLS)), result.levels)

    def test_export_hazard_to_csv(self):
        myhazards = list(query_v3.get_hazard_curves([loc.code for loc in LOCS[0:2]], VS30S, [HAZARD_MODEL_ID], IMTS))
        tmpcsv = io.StringIO()
        model_writer = csv.writer(tmpcsv)
        model_writer.writerows(model.HazardAggregation.to_csv(myhazards))

        tmpcsv.seek(0)
        self.assertTrue(next(tmpcsv).startswith('agg,imt,lat,lon,vs30,poe-0.001,poe-0.002'))
        self.assertEqual(len(list(tmpcsv)), len(myhazards))

    def test_export_hazard_to_dataframe_json(self):
        myhazards = list(query_v3.get_hazard_curves([loc.code for loc in LOCS[0:2]], VS30S, [HAZARD_MODEL_ID], IMTS))
        tmpcsv = io.StringIO()
        model_writer = csv.writer(tmpcsv)
        model_writer.writerows(model.HazardAggregation.to_csv(myhazards))


class CodedLocationResampling(unittest.TestCase):
    def test_get_nearest_hazard_for_an_arbitrary_location(self):
        gridloc = random.choice(LOCS)
        print(f'gridloc {gridloc}')

        off_lat = round(gridloc.lat + random.randint(0, 9) * 0.01, 3)
        off_lon = round(gridloc.lon + random.randint(0, 9) * 0.01, 3)
        somewhere_off_grid = CodedLocation(off_lat, off_lon, 0.001)

        nearest_grid = somewhere_off_grid.downsample(0.2)

        print(f'somewhere_off_grid {somewhere_off_grid}')
        print(f'nearest_grid {nearest_grid}')

        self.assertEqual(gridloc, nearest_grid.resample(0.001))
        self.assertEqual(gridloc, nearest_grid.downsample(0.001))
        self.assertEqual(gridloc.code, nearest_grid.downsample(0.001).code)

        self.assertEqual(gridloc, CodedLocation(nearest_grid.lat, nearest_grid.lon, 0.001))
        self.assertTrue(CodedLocation(nearest_grid.lat, nearest_grid.lon, 0.001) in LOCS)


@unittest.skip("not implemented")
class HighLevelRealizations(unittest.TestCase):
    def setUp(self):
        pass

    def test_get_the_realisations_for_hazard_location_from_hazard_set(self):
        assert 0


@unittest.skip("not implemented")
class HighLevelDeAggregations(unittest.TestCase):
    def setUp(self):
        pass

    def test_get_the_realisations_for_hazard_location_from_hazard_set(self):
        assert 0


toshi_hazard_utils/hazard/hazard.py
"""Helpers for obtaining hazard data from Toshi Hazard Store."""
from dataclasses import dataclass, field
from typing import Dict, Iterable, Set  # List, Set, Tuple

from nzshm_common.location.code_location import CodedLocation
from toshi_hazard_store import query_v3


@dataclass
class ToshiHazardReport:
    loc: CodedLocation
    model_id: str = field(default_factory=str)
    vs30s: Set[float] = field(default_factory=set)
    imts: Set[str] = field(default_factory=set)
    aggs: Set[str] = field(default_factory=set)
    levels: Set[Iterable[float]] = field(default_factory=set)


def hazard_report(hazard_model_ids: Iterable[str], locations: Iterable[CodedLocation]) -> Iterable[ToshiHazardReport]:
    """Get the availabe hazard attributes as ToshiHazardReport objects generator."""

    loc_reports: Dict[CodedLocation, ToshiHazardReport] = {}
    for res in query_v3.get_hazard_curves(locations, vs30s=[], hazard_model_ids=hazard_model_ids, imts=[]):
        loc_hmi = f'{CodedLocation(res.lat, res.lon, 0.001).code}:{res.hazard_model_id}'

        if loc_hmi not in loc_reports.keys():
            loc_reports[loc_hmi] = ToshiHazardReport(
                CodedLocation(res.lat, res.lon, 0.001)
            )  # , {[0]}, {[]}, {[]}, {[]})  # noqa

        report = loc_reports[loc_hmi]
        report.model_id = res.hazard_model_id
        report.imts.add(res.imt)
        report.vs30s.add(res.vs30)
        report.aggs.add(res.agg)
        for v in res.values:
            report.levels.add(v.lvl)

    for loc, report in loc_reports.items():
        yield report


toshi_hazard_utils/cli.py
"""Console script for toshi_hazard_utils."""

import csv
import io
import logging
import sys

import click
import toml
from click_option_group import MutuallyExclusiveOptionGroup, optgroup

try:
    import pandas as pd
except ImportError:
    print("WARNING: export to json (pandas dataframe) required uses the optional dependency - pandas.")

from nzshm_common.grids import RegionGrid
from nzshm_common.location import CodedLocation
from toshi_hazard_store import model, query_v3

from toshi_hazard_utils.hazard import hazard_report

log = logging.getLogger()
logging.basicConfig(level=logging.INFO)
logging.getLogger('nshm_toshi_client.toshi_client_base').setLevel(logging.INFO)
logging.getLogger('urllib3').setLevel(logging.INFO)
logging.getLogger('botocore').setLevel(logging.INFO)
logging.getLogger('pynamodb').setLevel(logging.INFO)
logging.getLogger('toshi_hazard_post').setLevel(logging.DEBUG)
logging.getLogger('toshi_hazard_store').setLevel(logging.INFO)
logging.getLogger('gql.transport.requests').setLevel(logging.WARN)

formatter = logging.Formatter(fmt='%(asctime)s %(levelname)-8s %(message)s', datefmt='%Y-%m-%d %H:%M:%S')
screen_handler = logging.StreamHandler(stream=sys.stdout)
screen_handler.setFormatter(formatter)
log.addHandler(screen_handler)


@click.group()
def thu():
    pass


@thu.command(name='report')
@click.option('-H', '--hazard_model_ids', help='comma-delimted list of hazard model ids.')
@optgroup.group('Site sources', cls=MutuallyExclusiveOptionGroup, help='The sites')
@optgroup.option('-S', '--sites', help='comma-delimited list of location codes.')
@optgroup.option('-L', '--locations_id', help='Extract sites from site list ENUM.')
@click.option('-c', '--config', type=click.Path(exists=True))  # help="path to a valid THU configuration file."
@click.option('-rs', '--resample', type=click.Choice(choices=['0.1', '0.2']), default=None)
@click.option('-v', '--verbose', is_flag=True)
def cli_hazard_report(hazard_model_ids, sites, site_lists, config, resample, verbose):
    """Gather information on available hazard curves for a given site or sites."""

    if config:
        conf = toml.load(config)
        if verbose:
            click.echo(f"using settings in {config} for report")
        sites = sites or conf.get('sites')
        hazard_model_ids = hazard_model_ids or conf['report'].get('hazard_model_ids')

    # parse sites into locations
    locations = []
    for site in sites.split(','):
        loc = CodedLocation(*[float(s) for s in site.split('~')], resolution=0.001)
        loc = loc.resample(float(resample)) if resample else loc
        locations.append(loc.resample(0.001).code)

    # parse hazard_model_ids from option
    hmids = []
    for hmid in hazard_model_ids.split(','):
        hmids.append(hmid.strip())

    if verbose:
        click.echo(f"locations {locations}")

    # Crazy - need to wrap generator in list() until we've squashed all the stray 'print' statements
    myhazards = list(hazard_report(hazard_model_ids=hmids, locations=locations))

    DESC = "Hazard Aggregation report"
    # INFO = "Gather information on available hazard curves for a given site or sites."
    click.echo("")
    click.echo(DESC)
    click.echo("=" * len(DESC))
    for haz in myhazards:
        click.echo(haz)


@thu.command(name='export')
@click.option('-H', '--hazard_model_ids', help='comma-delimted list of hazard model ids.')
@optgroup.group('Site sources', cls=MutuallyExclusiveOptionGroup, help='The sites ')
@optgroup.option('-S', '--sites', help='comma-delimited list of location codes.')
@optgroup.option('-L', '--site-lists', help='Extract sites from site list ENUM.')
# @click.option('-S', '--sites', help='comma-delimited list of location codes.')
# @click.option('-L', '--location_list', help='Extract sites from site list ENUM. ')
@click.option('-I', '--imts', help='comma-delimited list of imts.')
@click.option('-A', '--aggs', help='comma-delimited list of aggs.')
@click.option('-V', '--vs30s', help='comma-delimited list of vs30s.')
@click.option('-c', '--config', type=click.Path(exists=True))  # help="path to a valid THU configuration file."
@click.option('-rs', '--resample', type=click.Choice(choices=['0.1', '0.2']), default=None)
@click.option('-v', '--verbose', is_flag=True)
@click.option('-o', '--output', type=click.File('w'), default='-')
@click.option('-f', '--format', type=click.Choice(choices=['csv', 'json']), default='csv')
def cli_hazard_export(
    hazard_model_ids, sites, site_lists, imts, aggs, vs30s, config, resample, verbose, output, format
):
    """Export hazard curves for a given set of arguments."""

    sites = sites.split(',') if sites else None
    site_lists = site_lists.split(',') if site_lists else None

    if sites and site_lists:
        raise ValueError("One of sites or site_lists arguments must be supplied.")

    hazard_model_ids = hazard_model_ids.split(',') if hazard_model_ids else None
    imts = imts.split(',') if imts else None
    vs30s = vs30s.split(',') if vs30s else None
    aggs = aggs.split(',') if aggs else None

    if config:
        conf = toml.load(config)
        if verbose:
            click.echo(f"using settings in {config} for export")
        sites = sites or conf.get('sites', '').split(',')
        site_lists = site_lists or conf.get('site_lists', [])
        hazard_model_ids = hazard_model_ids or conf.get('hazard_model_ids')
        imts = imts or conf.get('imts')
        vs30s = vs30s or conf.get('vs30s')
        aggs = aggs or conf.get('aggs')

    # parse sites into locations
    locations = []

    if site_lists:
        # Get locations from the supplied enum(s)
        for site_list in site_lists:
            grid = RegionGrid[site_list]
            grid_locs = grid.load()
            for gloc in grid_locs:
                loc = CodedLocation(*gloc, resolution=0.001)
                loc = loc.resample(float(resample)) if resample else loc
                locations.append(loc.resample(0.001).code)
    else:
        # convert locations fomr sites
        for site in sites:
            loc = CodedLocation(*[float(s) for s in site.split('~')], resolution=0.001)
            loc = loc.resample(float(resample)) if resample else loc
            locations.append(loc.resample(0.001).code)

    if verbose:
        click.echo(f"{sites} {hazard_model_ids} {imts} {vs30s} {format}")
    if verbose:
        click.echo(output)

    haggs = query_v3.get_hazard_curves(locations, vs30s, hazard_model_ids, imts=imts, aggs=aggs)

    if format == 'csv':
        model_writer = csv.writer(output)
        model_writer.writerows(list(model.HazardAggregation.to_csv(haggs)))
        return

    if format == 'json':
        # first write to an in memory csv
        tmpcsv = io.StringIO()
        model_writer = csv.writer(tmpcsv)
        model_writer.writerows(list(model.HazardAggregation.to_csv(haggs)))
        tmpcsv.seek(0)
        # now build a dataframe
        df = pd.read_csv(tmpcsv)
        if verbose:
            click.echo(df)
        df.to_json(output)
        return


if __name__ == "__main__":
    thu()  # pragma: no cover
